Boot upgrade without jx namespace in kube config

### Summary 

When executing `jx upgrade boot` with kube config NOT pointing to the `dev` (e.g., `jx`) Namespace, the output is as follows.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x22766da]

goroutine 1 [running]:
github.com/jenkins-x/jx/pkg/gits.(*GitCLI).CreateAuthenticatedURL(0xc0009aea98, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /workspace/source/pkg/gits/git_cli.go:581 +0x6a
github.com/jenkins-x/jx/pkg/cmd/upgrade.(*UpgradeBootOptions).cloneDevEnv(0xc00054ae40, 0x0, 0x0)
        /workspace/source/pkg/cmd/upgrade/upgrade_boot.go:488 +0x1b5
github.com/jenkins-x/jx/pkg/cmd/upgrade.(*UpgradeBootOptions).Run(0xc00054ae40, 0x0, 0x0)
        /workspace/source/pkg/cmd/upgrade/upgrade_boot.go:83 +0xad3
github.com/jenkins-x/jx/pkg/cmd/upgrade.NewCmdUpgradeBoot.func1(0xc000806280, 0xb162868, 0x0, 0x0)
        /workspace/source/pkg/cmd/upgrade/upgrade_boot.go:64 +0x6f
github.com/spf13/cobra.(*Command).execute(0xc000806280, 0xb162868, 0x0, 0x0, 0xc000806280, 0xb162868)
        /workspace/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830 +0x2ae
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000a8000, 0xc000309c00, 0x6608560, 0xc00000e010)
        /workspace/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fc
github.com/spf13/cobra.(*Command).Execute(...)
        /workspace/go/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
github.com/jenkins-x/jx/cmd/jx/app.Run(0x0, 0x0, 0x0, 0xc0006f9f88, 0xc0000a0058)
        /workspace/source/cmd/jx/app/jx.go:19 +0x185
main.main()
        /workspace/source/cmd/jx/jx.go:11 +0x32
```

### Steps to reproduce the behavior

Change the namespace in kube config to something other than Jenkins X `dev` namespace. Execute `jx upgrade boot`

### Expected behavior

An user-friendly message that tells that the problem is that kube config does not point to the `dev` Namespace.

### Actual behavior

`panic`

### Jx version

```
NAME               VERSION
jx                 2.0.1126
Kubernetes cluster v1.14.8-gke.17
kubectl            v1.17.1
git                2.23.0
Operating System   Mac OS X 10.12.6 build 16G2136
```

### Jenkins type

- [X] Serverless Jenkins X Pipelines (Tekton + Prow)
- [ ] Classic Jenkins

### Kubernetes cluster

Any

### Operating system / Environment

Mac OS

--- a/pkg/cmd/opts/git.go
+++ b/pkg/cmd/opts/git.go
@@ -350,6 +350,9 @@ func (o *CommonOptions) getAuthConfig() (*auth.AuthConfig, error) {
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to create the git auth config service")
 	}
+	if authConfigSvc == nil {
+		return nil, errors.New("empty auth config")
+	}
 	authConfig := authConfigSvc.Config()
 	if authConfig == nil {
 		return nil, errors.New("empty Git config")

--- a/pkg/cmd/upgrade/upgrade_boot.go
+++ b/pkg/cmd/upgrade/upgrade_boot.go
@@ -449,13 +449,15 @@ func (o *UpgradeBootOptions) cherryPickCommits(cloneDir, fromSha, toSha string) 
 
 func (o *UpgradeBootOptions) setupGitConfig(dir string) error {
 	jxClient, devNs, err := o.JXClientAndDevNamespace()
+	if err != nil {
+		return errors.Wrap(err, "failed to create/get jx client and dev namespace")
+	}
 	devEnv, err := kube.GetDevEnvironment(jxClient, devNs)
 	if err != nil {
 		return errors.Wrapf(err, "failed to get dev environment in namespace %s", devNs)
 	}
 	username := devEnv.Spec.TeamSettings.PipelineUsername
 	email := devEnv.Spec.TeamSettings.PipelineUserEmail
-
 	err = o.Git().SetUsername(dir, username)
 	if err != nil {
 		return errors.Wrapf(err, "failed to set username %s", username)
@@ -521,9 +523,15 @@ func (o *UpgradeBootOptions) deleteLocalBranch(branch string) error {
 
 func (o *UpgradeBootOptions) cloneDevEnv() error {
 	jxClient, devNs, err := o.JXClientAndDevNamespace()
+	if err != nil {
+		return errors.Wrap(err, "failed to create/get jx client and dev namespace")
+	}
 	devEnv, err := kube.GetDevEnvironment(jxClient, devNs)
 	if err != nil {
 		return errors.Wrapf(err, "failed to get dev environment in namespace %s", devNs)
+	}
+	if devEnv == nil {
+		return errors.Wrapf(errors.New("Cannot find environment with dev selector"), "No development environment found for namespace %s", devNs)
 	}
 	devEnvURL := devEnv.Spec.Source.URL
 
@@ -537,6 +545,9 @@ func (o *UpgradeBootOptions) cloneDevEnv() error {
 	}
 
 	gitInfo, err := gits.ParseGitURL(devEnvURL)
+	if err != nil {
+		return errors.Wrapf(err, "failed to parse url %s", devEnvURL)
+	}
 	_, userAuth, err := o.GetPipelineGitAuthForRepo(gitInfo)
 	if err != nil {
 		return errors.Wrap(err, "failed to get pipeline user auth")

--- a/pkg/cmd/upgrade/upgrade_boot_test.go
+++ b/pkg/cmd/upgrade/upgrade_boot_test.go
@@ -14,10 +14,20 @@ import (
 	"github.com/jenkins-x/jx/pkg/log"
 
 	"github.com/jenkins-x/jx/pkg/cmd/opts"
+	"github.com/jenkins-x/jx/pkg/cmd/testhelpers"
 	"github.com/jenkins-x/jx/pkg/config"
 	"github.com/jenkins-x/jx/pkg/gits"
+	"github.com/jenkins-x/jx/pkg/kube"
 	"github.com/jenkins-x/jx/pkg/util"
 	"github.com/jenkins-x/jx/pkg/versionstream"
+
+	jenkinsv1 "github.com/jenkins-x/jx/pkg/apis/jenkins.io/v1"
+	clientfake "github.com/jenkins-x/jx/pkg/cmd/clients/fake"
+	helm_test "github.com/jenkins-x/jx/pkg/helm/mocks"
+	resources_test "github.com/jenkins-x/jx/pkg/kube/resources/mocks"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
@@ -32,12 +42,71 @@ type TestUpgradeBootOptions struct {
 	Dir string
 }
 
-func (o *TestUpgradeBootOptions) setup(bootRequirements string) {
+func (o *TestUpgradeBootOptions) setup(bootRequirements, previewEnvironmentName, sourceURL, namespace string) {
 	dir := bootRequirements
+	// Not setting it to "dev" will overwrite this preview environment with a new one
+	devEnv := kube.NewPreviewEnvironment(previewEnvironmentName)
+	devEnv.Spec.Source.URL = sourceURL
+	commonOpts := &opts.CommonOptions{}
+	commonOpts.SetDevNamespace(namespace)
+
+	factory := clientfake.NewFakeFactory()
+	user := jenkinsv1.User{
+		ObjectMeta: v1.ObjectMeta{
+			Namespace: "jx",
+			Name:      "test-user",
+		},
+		Spec: jenkinsv1.UserDetails{
+			Name:     "Test",
+			Email:    "test@test.com",
+			Accounts: make([]jenkinsv1.AccountReference, 0),
+		},
+	}
+
+	testhelpers.ConfigureTestOptionsWithResources(commonOpts,
+		[]runtime.Object{},
+		[]runtime.Object{&user, devEnv},
+		&gits.GitFake{CurrentBranch: "job"},
+		&gits.FakeProvider{},
+		helm_test.NewMockHelmer(),
+		resources_test.NewMockInstaller(),
+	)
+
+	commonOpts.SetFactory(factory)
 
 	o.UpgradeBootOptions = UpgradeBootOptions{
-		CommonOptions: &opts.CommonOptions{},
+		CommonOptions: commonOpts,
 		Dir:           dir,
+	}
+}
+
+var clonetests = []struct {
+	desc     string
+	pEnvName string
+	sURL     string
+	ns       string
+	success  bool
+}{
+	{"Test should pass - success", "dev", "https://gitlab.com/ankitm123/environment-mikros-cluster-dev.git", "jx", true},
+	{"non-dev preview environment - fail", "testing", "https://gitlab.com/ankitm123/environment-mikros-cluster-dev.git", "jx", false},
+	{"non-dev ns - fail", "dev", "https://gitlab.com/ankitm123/environment-mikros-cluster-dev.git", "no-jx", false},
+	{"missing application url - fail", "dev", "", "jx", false},
+}
+
+func TestCloneDevEnvironment(t *testing.T) {
+	t.Parallel()
+	for _, tt := range clonetests {
+		t.Run(tt.desc, func(t *testing.T) {
+			o := TestUpgradeBootOptions{}
+			o.setup(defaultBootRequirements, tt.pEnvName, tt.sURL, tt.ns)
+			err := o.cloneDevEnv()
+			if tt.success {
+				require.NoError(t, err, "cloned dev")
+			} else {
+				require.Error(t, err, "could not clone dev")
+			}
+
+		})
 	}
 }
 
@@ -45,7 +114,7 @@ func TestDetermineBootConfigURL(t *testing.T) {
 	t.Parallel()
 
 	o := TestUpgradeBootOptions{}
-	o.setup(defaultBootRequirements)
+	o.setup(defaultBootRequirements, "", "", "")
 
 	requirements, _, err := config.LoadRequirementsConfig(o.UpgradeBootOptions.Dir)
 	require.NoError(t, err, "could not get requirements file")
@@ -60,7 +129,7 @@ func TestRequirementsVersionStream(t *testing.T) {
 	t.Parallel()
 
 	o := TestUpgradeBootOptions{}
-	o.setup(defaultBootRequirements)
+	o.setup(defaultBootRequirements, "", "", "")
 
 	requirements, _, err := config.LoadRequirementsConfig(o.UpgradeBootOptions.Dir)
 	require.NoError(t, err, "could not get requirements file")
@@ -74,7 +143,7 @@ func TestUpdateVersionStreamRef(t *testing.T) {
 	t.Parallel()
 
 	o := TestUpgradeBootOptions{}
-	o.setup(defaultBootRequirements)
+	o.setup(defaultBootRequirements, "", "", "")
 
 	tmpDir := o.createTmpRequirements(t)
 	defer func() {
@@ -97,7 +166,7 @@ func TestUpdatePipelineBuilderImage(t *testing.T) {
 	t.Parallel()
 
 	o := TestUpgradeBootOptions{}
-	o.setup(defaultBootRequirements)
+	o.setup(defaultBootRequirements, "", "", "")
 
 	tmpDir, err := ioutil.TempDir("", "")
 	defer func() {
@@ -126,7 +195,7 @@ func TestUpdateTemplateBuilderImage(t *testing.T) {
 	t.Parallel()
 
 	o := TestUpgradeBootOptions{}
-	o.setup(defaultBootRequirements)
+	o.setup(defaultBootRequirements, "", "", "")
 
 	tmpDir, err := ioutil.TempDir("", "")
 	defer func() {
@@ -170,7 +239,7 @@ func TestDetermineBootConfigURLAlternative(t *testing.T) {
 	t.Parallel()
 
 	o := TestUpgradeBootOptions{}
-	o.setup(alternativeBootRequirements)
+	o.setup(alternativeBootRequirements, "", "", "")
 
 	requirements, _, err := config.LoadRequirementsConfig(o.UpgradeBootOptions.Dir)
 	require.NoError(t, err, "could not get requirements file")
@@ -185,7 +254,7 @@ func TestRequirementsVersionStreamAlternative(t *testing.T) {
 	t.Parallel()
 
 	o := TestUpgradeBootOptions{}
-	o.setup(alternativeBootRequirements)
+	o.setup(alternativeBootRequirements, "", "", "")
 
 	requirements, _, err := config.LoadRequirementsConfig(o.UpgradeBootOptions.Dir)
 	require.NoError(t, err, "could not get requirements file")
@@ -199,7 +268,7 @@ func TestUpdateVersionStreamRefAlternative(t *testing.T) {
 	t.Parallel()
 
 	o := TestUpgradeBootOptions{}
-	o.setup(alternativeBootRequirements)
+	o.setup(alternativeBootRequirements, "", "", "")
 
 	tmpDir := o.createTmpRequirements(t)
 	defer func() {

